samtranslator/model/sam_resources.py

```diff
@@ -2300,9 +2302,55 @@ def _construct_ddb_datasources(
                 cfn_datasource.Description = cast(PassThrough, ddb_datasource.Description)
 
             cfn_datasource.DynamoDBConfig = self._parse_ddb_config(ddb_datasource)
-            datasources.append(cfn_datasource)
 
-        return datasources
+            cfn_datasource.ServiceRoleArn, role, connector_resources = self.parse_datasource_role(
+                ddb_datasource, cfn_datasource.get_runtime_attr("arn"), relative_id, kwargs
+            )
+
+            if role:
+                resources.append(role)
+
+            if connector_resources:
+                resources.extend(connector_resources)
```

Comment:
I assume no need to return `role` and `connector_resources` separately if they're all added to an opaque `resources` anyway?

Reply:
Yea exactly. So we'll move our `ServiceRoleArn` check back into the loop, and then only execute this function when we know we need to generate `role` and `connector_resources`. Should make this much more intuitive